design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+      "source": {
+        "packaging": "zip",
+        "sourceHash": "a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57",
+        "sourcePath": "asset.a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57"
+      },
+      "destinations": [
+        {
+          "bucketName": "aws-cdk-files-2222222222US-us-east-1",
+          "objectKey": "a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57",
+          "assumeRoleArn": "arn:aws:iam::2222222222US:role/aws-cdk-publish-2222222222US-us-east-1"
+        },
+        {
+          "bucketName": "aws-cdk-files-3333333333EU-us-west-2",
+          "objectKey": "a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57",
+          "assumeRoleArn": "arn:aws:iam::3333333333EU:role/aws-cdk-publish-3333333333EU-eu-west-2"
+        }
+      ]
+    }
+  }
+}
+```
+
+It lists the two assets (one file asset and one image asset) and, for each, it lists the publishing locations which include repository, key and publishing IAM role to assume.
+
+### Mutation
+
+The first stage in our pipeline is the mutation stage. This stage will include a single CodeBuild action that will simply execute:
+
+```console
+$ cdk deploy pipeline-*
+```
+
+This will update all the stacks with names that begin with "pipeline-". Namely, it includes the pipeline stack itself, and the auxiliary stacks that include the pipeline's regional replication buckets. Since we are using `cdk deploy` here, we can technically deploy any stack to any environment in this stage because `cdk deploy` can assume the deployment role in any of the environments which trust the deployment account.
+
+Notice that this stage happens **before** the publish stage. This is because the structure of the publish stage is dependent on which assets the app uses (we synthesize an action per asset for maximal parallelism), so we want to make sure the pipeline will be updated before publishing.
+
+> NOTE: if there is a constraint that does not allow the pipeline to be updated before publishing, it simply means that we have to publish all assets from a single action (`cdk-assets publish cdk.out`).
+
+In our example, there are 3 pipeline stacks:
+
+1. `pipeline-main`: contains the pipeline resource, the main artifacts bucket and other related resources
+2. `pipeline-us-east-1`: contains the pipeline replication bucket and key for us-east-1
+3. `pipeline-eu-west-2`: contains the pipeline replication bucket and key for eu-west-2
+
+Once we deploy these stacks, our pipeline will be ready to deploy the rest of our app.
+
+### Publishing
+
+The next stage in the process is the publishing stage.
+
+In our example, this stage will consist of two CodeBuild actions that will run the following commands concurrently. These command will package & upload the asset to all the environments. It will consult `assets.json` to determine the exact location into which to publish each asset and which cross-account role to assume.
```

Comment:
two because there are two assets? How did the `pipeline-` stack know of this?

When another asset is added as part of the service stack, does `cdk synth` automatically add another CodeBuild action? How is this information transmitted from the service stack to the synthesis of the pipeline stack? via `assets.json`?

Reply:
The synthesized "pipeline-main" stack will change when new assets are consumed by the app. The CI/CD APIs will synthesize the asset publishing stage based on the information included in the app.